utils/s2n_map.h

```diff
@@ -30,3 +31,9 @@ S2N_RESULT s2n_map_complete(struct s2n_map *map);
 S2N_RESULT s2n_map_unlock(struct s2n_map *map);
 S2N_RESULT s2n_map_lookup(const struct s2n_map *map, struct s2n_blob *key, struct s2n_blob *value, bool *key_found);
 S2N_RESULT s2n_map_free(struct s2n_map *map);
+S2N_RESULT s2n_map_size(struct s2n_map *map, uint32_t *size);
+
+S2N_RESULT s2n_map_iterator_new(struct s2n_map *map, struct s2n_map_iterator **iter);
```

Comment:
It's very likely more efficient to make this s2n_map_iterator_init instead. You'd need to move the definition of the iterator to the non-internal header, but that's probably fine. I'm not convinced this is worth an alloc.

Reply:
Hmmm or do you even need an iterator? If all you track is the index, you could just need a method to get a specific index. That + your size method would cover the usecase.

Reply:
Will do on the `interator_init` thing!

And I guess that's more of a stylistic thing. 
> you could just need a method to get a specific index

This feels like it would go against the goal with restricting map internals, since it would be permitting arbitrary access to any slot in the hash table. The nice thing about the current iterator interface is that we can at least ensure that all of the values yielded will always be valid elements.